### PR TITLE
STITCH-3383 add client sdk FunctionCredential

### PR DESCRIPTION
**Pull request by jsflax**

### CODE REVIEW

File: packages/core/admin-client/src/authProviders/ProviderConfigs.ts [hunk 1]
```diff
@@ -85,5 +100,7 @@ export {
   UserpassProviderConfig, 
   UserpassProvider, 
   CustomProviderConfig, 
-  CustomProvider
+  CustomProvider,
+  FunctionProviderConfig,
```
Comment (mburdette) [inline packages/core/admin-client/src/authProviders/ProviderConfigs.ts]: [nit] I think in other places we are calling this `CustomFunction` not sure if its worth making the change here as well?

File: packages/core/admin-client/src/index.ts [hunk 2]
```diff
@@ -22,7 +22,9 @@ import {
   AnonProviderConfig, 
   ApiKey,
   CustomProvider, 
-  CustomProviderConfig, 
+  CustomProviderConfig,
+  FunctionProvider,
+  FunctionProviderConfig,
```
Comment (mburdette) [inline packages/core/admin-client/src/index.ts]: and here

File: packages/core/admin-client/src/index.ts [hunk 3]
```diff
@@ -110,6 +112,8 @@ export {
   ServicesResource,
   CustomProviderConfig, 
   CustomProvider,
+  FunctionProvider,
+  FunctionProviderConfig,
```
Comment (mburdette) [inline packages/core/admin-client/src/index.ts]: and here

File: packages/react-native/coretest/__tests__/StitchAppClientIntTests.ts [hunk 4]
```diff
@@ -520,6 +524,36 @@ describe("StitchAppClient", () => {
...
+      expect(user.id).toBeDefined();
+      expect(user.identities[0].id).toBeDefined();
+      expect("foo").toEqual(user.identities[0].id);
+      expect(FunctionAuthProvider.DEFAULT_NAME).toEqual(user.loggedInProviderName);
+      expect(FunctionAuthProvider.TYPE).toEqual(user.loggedInProviderType);
+      expect(UserType.Normal).toEqual(user.userType);
+      expect(FunctionAuthProvider.TYPE).toEqual(user.identities[0].providerType);
+      expect(client.auth.isLoggedIn).toBeTruthy();
+  });
+
```
Comment (mburdette) [inline packages/react-native/coretest/__tests__/StitchAppClientIntTests.ts]: can we also add some fail tests as well as tests that actually use whats in the payload?

File: packages/server/coretest/__tests__/StitchAppClientIntTests.ts [hunk 5]
```diff
@@ -518,6 +522,36 @@ describe("StitchAppClient", () => {
     expect(client.auth.listUsers().length).toBe(2);
   });
 
+  it("should authenticate using a custom function", async () => {
```
Comment (mburdette) [inline packages/server/coretest/__tests__/StitchAppClientIntTests.ts]: can we also add some fail tests as well as tests that actually use whats in the payload?

File: packages/server/coretest/__tests__/StitchAppClientIntTests.ts [hunk 6]
```diff
@@ -518,6 +522,36 @@ describe("StitchAppClient", () => {
...
+
+      const client = harness.getAppClient(appResponse);
+
+      const user = await client.auth.loginWithCredential(new FunctionCredential({"id": "123abc"}));
+
+      expect(user).toBeDefined();
+
+      expect(user.id).toBeDefined();
+      expect(user.identities[0].id).toBeDefined();
+      expect("foo").toEqual(user.identities[0].id);
```
Comment (mburdette) [inline packages/server/coretest/__tests__/StitchAppClientIntTests.ts]: [nit] if you gonna use a string in two places use a `const`

File: packages/react-native/coretest/__tests__/StitchAppClientIntTests.ts [hunk 7]
```diff
@@ -520,6 +524,36 @@ describe("StitchAppClient", () => {
...
+      const client = await harness.getAppClient(appResponse);
+
+      const user = await client.auth.loginWithCredential(new FunctionCredential({"id": "123abc"}));
+
+      expect(user).toBeDefined();
+
+      expect(user.id).toBeDefined();
+      expect(user.identities[0].id).toBeDefined();
+      expect("foo").toEqual(user.identities[0].id);
```
Comment (mburdette) [inline packages/react-native/coretest/__tests__/StitchAppClientIntTests.ts]: [nit] if you gonna use a string in two places use a `const`

File: packages/browser/coretest/__tests__/StitchAppClientIntTests.ts [hunk 8]
```diff
@@ -520,6 +524,36 @@ describe("StitchAppClient", () => {
...
+
+      const client = harness.getAppClient(appResponse);
+
+      const user = await client.auth.loginWithCredential(new FunctionCredential({"id": "123abc"}));
+
+      expect(user).toBeDefined();
+
+      expect(user.id).toBeDefined();
+      expect(user.identities[0].id).toBeDefined();
+      expect("foo").toEqual(user.identities[0].id);
```
Comment (mburdette) [inline packages/browser/coretest/__tests__/StitchAppClientIntTests.ts]: [nit] if you gonna use a string in two places use a `const`

File: packages/browser/coretest/__tests__/StitchAppClientIntTests.ts [hunk 9]
```diff
@@ -520,6 +524,36 @@ describe("StitchAppClient", () => {
...
+      expect(FunctionAuthProvider.DEFAULT_NAME).toEqual(user.loggedInProviderName);
+      expect(FunctionAuthProvider.TYPE).toEqual(user.loggedInProviderType);
+      expect(UserType.Normal).toEqual(user.userType);
+      expect(FunctionAuthProvider.TYPE).toEqual(user.identities[0].providerType);
+      expect(client.auth.isLoggedIn).toBeTruthy();
+  });
+
```
Comment (mburdette) [inline packages/browser/coretest/__tests__/StitchAppClientIntTests.ts]: can we also add some fail tests as well as tests that actually use whats in the payload?

**Comment by jsflax** [comment]
@mburdette PTAL

**Pull request merged**